arbitrum-client/src/client/event_indexing.rs

```diff
@@ -0,0 +1,105 @@
+use constants::{Scalar, MERKLE_HEIGHT};
+use ethers::types::TxHash;
+use num_bigint::BigUint;
+
+use crate::{
+    abi::{NodeChangedFilter, WalletUpdatedFilter},
+    constants::DEFAULT_AUTHENTICATION_PATH,
+    errors::ArbitrumClientError,
+    helpers::hash_scalar,
+    serde_def_types::SerdeScalarField,
+};
+
+use super::ArbitrumClient;
+
+impl ArbitrumClient {
+    /// Return the hash of the transaction that last indexed secret shares for
+    /// the given public blinder share
+    ///
+    /// Returns `None` if the public blinder share has not been used
+    pub async fn get_public_blinder_tx(
+        &self,
+        public_blinder_share: Scalar,
+    ) -> Result<Option<TxHash>, ArbitrumClientError> {
+        let public_blinder_share_hash = hash_scalar(public_blinder_share)?;
+        let events = self
+            .darkpool_event_source
+            .event::<WalletUpdatedFilter>()
+            .topic1(public_blinder_share_hash)
+            .from_block(self.deploy_block)
+            .query_with_meta()
+            .await
+            .map_err(|e| ArbitrumClientError::EventQuerying(e.to_string()))?;
+
+        Ok(events.last().map(|(_, meta)| meta.transaction_hash))
+    }
+
+    /// Searches on-chain state for the insertion of the given wallet, then
+    /// finds the most recent updates of the path's siblings and creates a
+    /// Merkle authentication path
+    pub async fn find_merkle_authentication_path(
+        &self,
+        commitment: Scalar,
+    ) -> Result<MerkleAuthenticationPath, ArbitrumClientError> {
+        let leaf_index = BigUint::from(self.find_commitment_in_state(commitment).await?);
+
+        // Construct a set that holds pairs of (depth, index) values in the
+        // authentication path; i.e. the tree coordinates of the sibling nodes
+        // in the authentication path
+        let mut authentication_path_coords =
+            MerkleAuthenticationPath::construct_path_coords(leaf_index.clone(), MERKLE_HEIGHT);
+
+        // For each coordinate in the authentication path,
+        // find the last value it was updated to
+        let mut path = *DEFAULT_AUTHENTICATION_PATH;
+        for coords in authentication_path_coords {
+            let events = self
+                .darkpool_event_source
+                .event::<NodeChangedFilter>()
+                .topic1(coords.height.into())
+                .topic2(coords.index.into())
+                .from_block(self.deploy_block)
+                .query()
+                .await
+                .map_err(|e| ArbitrumClientError::EventQuerying(e.to_string()))?;
+
+            let value = events
+                .last()
+                .map(|event| {
+                    postcard::from_bytes::<SerdeScalarField>(&event.new_value)
+                        .map_err(|e| ArbitrumClientError::Serde(e.to_string()))?
+                        .0 // Scalar
+                })
+                .ok_or(ArbitrumClientError::NodeChangeNotFound)?;
```

Comment:
This shouldn't error, it should just do nothing, i.e. leave the default value in the authentication path.